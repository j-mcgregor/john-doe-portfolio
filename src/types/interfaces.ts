export interface NavbarProps {
    brand: string | JSX.Element
}

export interface PRISMIC_RichTextType {
    type: string
    text: string
    spans: any[]
}

export interface PRISMIC_Image {
    dimensions: {
        width: number
        height: number
    }
    alt?: string | null
    copyright: string | null
    url: string
}

export interface PRISMIC_ExperienceType {
    job_title: string
    date_from: string
    date_to?: string
    present?: boolean
    company: string
    city: string
    description: PRISMIC_RichTextType[]
}

export interface PRISMIC_landing_node {
    node: {
        primary_text: PRISMIC_RichTextType[]
        secondary_text: PRISMIC_RichTextType[]
        background_image: PRISMIC_Image
        about: PRISMIC_RichTextType[]
        experience: PRISMIC_ExperienceType[]
    }
}

export interface IndexPageProps {
    data: {
        prismic: {
            allLandings: {
                edges: PRISMIC_landing_node[]
            }
        }
        file: {
            childImageSharp: {
                fixed: {
                    height: number
                    width: number
                    base64: string
                    src: string
                    srcSet: string
                }
            }
        }
    }
}

export interface HeroProps {
    primary: JSX.Element
    secondary: JSX.Element
    logo?: JSX.Element
    backgroundImage?: string
}

export interface StyledHeroProps {
    bgImageUrl?: string
}

export interface AboutProps {
    about: JSX.Element
    experience: JSX.Element[]
}

// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>/  FORMS   >>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

export interface TextInputProps {
    type: string
    placeholder: string
    label?: string
    name: string
    required: boolean
    classNames?: string[]
    value: string
    onChange: (e: React.FormEvent<HTMLInputElement>) => void
}

export interface TextAreaInputProps {
    placeholder: string
    label?: string
    name: string
    required: boolean
    value: string
    onChange: (e: React.FormEvent<HTMLTextAreaElement>) => void
}
