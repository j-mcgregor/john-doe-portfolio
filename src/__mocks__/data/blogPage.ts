import _ from 'lodash'
import {
    BlogPageProps,
    PRISMIC_PRIMARY_Image_with_Caption,
    PRISMIC_PRIMARY_Quote,
    PRISMIC_PRIMARY_Text,
} from '../../types/interfaces/blog'
import {
    PRISMIC_blog_node,
    PRISMIC_RichTextType,
    PRISMIC_MetaType,
} from '../../types/interfaces/prismic'

type ImageType = 'image_with_caption' | 'full-width-image' | 'emphasized'

type TextType = 'paragraph' | 'heading3'

export interface ImageSliceTypeProps {
    [index: string]: ImageType
}

export const imageSliceType: ImageSliceTypeProps = {
    '0': 'image_with_caption',
    '1': 'full-width-image',
    '2': 'emphasized',
}

export const makeTextType = (
    type: TextType,
    text: string
): PRISMIC_RichTextType => ({
    type,
    text,
    spans: [],
})

export const meta = (
    id: string,
    uid: string,
    type: string
): PRISMIC_MetaType => ({
    id,
    uid,
    type,
})

export const imageSlice = (
    imageType: number
): PRISMIC_PRIMARY_Image_with_Caption => ({
    primary: {
        caption: [makeTextType('paragraph', `caption-${imageType}`)],
        image: {
            dimensions: {
                width: 1200,
                height: 900,
            },
            alt: `alt-${imageType}`,
            copyright: null,
            url: 'http://fillmurray.com/300/300',
        },
    },
    label: `default`,
    type: imageSliceType[`${imageType}`],
})

export const quoteSlice = (text: string): PRISMIC_PRIMARY_Quote => ({
    primary: {
        quote: [makeTextType('paragraph', text)],
    },
    label: undefined,
    type: 'quote',
})

export const textSlice = (text: string): PRISMIC_PRIMARY_Text => ({
    primary: {
        text: [makeTextType('paragraph', text)],
    },
    label: undefined,
    type: 'text',
})

export const blogNodes: PRISMIC_blog_node[] = _.times(3, num => ({
    node: {
        date: `2020-0${num + 1}-01`,
        body: [
            textSlice('Lorem ipsum dolor sit amet'),
            imageSlice(num),
            quoteSlice('And I quoteth thus...'),
        ],
        _meta: meta(`id-${num + 1}`, `blog-${num + 1}`, 'blog_post'),
        title: [makeTextType('paragraph', `Post ${num + 1}`)],
    },
}))

export const blogPage: BlogPageProps = {
    data: {
        prismic: {
            allBlog_posts: {
                edges: blogNodes,
            },
        },
    },
}
