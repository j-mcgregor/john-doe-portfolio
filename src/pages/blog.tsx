import React from 'react'
import { graphql } from 'gatsby'

import Layout from '../components/Layout'
import SEO from '../components/SEO'
import { BlogPageProps } from '../types/interfaces/blog'
import BlogPosts from '../components/blog/BlogPosts'

export const query = graphql`
    query BlogPostsQuery {
        prismic {
            allBlog_posts {
                edges {
                    node {
                        title
                        date
                        body {
                            ... on PRISMIC_Blog_postBodyImage_with_caption {
                                primary {
                                    caption
                                    image
                                }
                                type
                            }
                            ... on PRISMIC_Blog_postBodyText {
                                primary {
                                    text
                                }
                                type
                            }
                        }
                        _meta {
                            id
                            type
                            uid
                        }
                    }
                }
            }
        }
    }
`

const BlogPage: React.FC<BlogPageProps> = ({
    data: {
        prismic: {
            allBlog_posts: { edges },
        },
    },
}) => {
    return (
        <Layout>
            <SEO title="Blog Post" />
            <BlogPosts posts={edges} />
        </Layout>
    )
}

export default BlogPage
