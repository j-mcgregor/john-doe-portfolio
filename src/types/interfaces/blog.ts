// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>  BLOG   >>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

import {
    PRISMIC_RichTextType,
    PRISMIC_Image,
    PRISMIC_blog_node,
    PRISMIC_blog_node_body,
} from './prismic'

export interface BlogBodyProps {
    __typename?: string
    type: string
    primary: {
        text?: PRISMIC_RichTextType[]
        caption?: string | null
        image?: PRISMIC_Image
    }
}

export interface BlogPageProps {
    data: {
        prismic: {
            allBlog_posts: {
                edges: PRISMIC_blog_node[]
            }
        }
    }
}

export interface PostSummaryProps {
    post: PRISMIC_blog_node_body
}

export interface BlogPostsProps {
    posts?: PRISMIC_blog_node[]
}
