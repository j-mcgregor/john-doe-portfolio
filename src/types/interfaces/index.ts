// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>/  INDEX   >>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

import { PRISMIC_landing_node } from './prismic'

export interface IndexPageProps {
    data: {
        prismic: {
            allLandings: {
                edges: PRISMIC_landing_node[]
            }
        }
        file: {
            childImageSharp: {
                fixed: {
                    height: number
                    width: number
                    base64: string
                    src: string
                    srcSet: string
                }
            }
        }
    }
}
