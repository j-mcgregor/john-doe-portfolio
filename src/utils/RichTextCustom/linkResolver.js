export const linkResolver = doc => {
    if (doc.type === 'page') {
        return `/${doc.uid}`
    }

    if (doc.type === 'blog_post') {
        return `/blog/${doc.uid}`
    }

    return '/'
}
