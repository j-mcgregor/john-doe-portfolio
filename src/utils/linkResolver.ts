interface LinkResolverProp {
    type: string
    uid: string
}

export const linkResolver = (doc: LinkResolverProp) => {
    if (doc.type === 'page') {
        return `/${doc.uid}`
    }

    return '/'
}
