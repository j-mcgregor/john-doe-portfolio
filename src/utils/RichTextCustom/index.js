import React from 'react'
import { RichText } from 'prismic-reactjs'
import { linkResolver } from './linkResolver'
import { htmlSerializer } from './htmlSerializer'

const RichTextCustom = ({ render }) => (
    <RichText
        render={render}
        linkResolver={linkResolver}
        htmlSerializer={htmlSerializer}
    />
)

export default RichTextCustom
