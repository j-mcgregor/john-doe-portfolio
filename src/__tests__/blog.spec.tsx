import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import Blog from '../pages/blog'
import { BlogPageProps } from '../types/interfaces/blog'
import { blogPage } from '../__mocks__/data/blogPage'
import * as Link from '../utils/RichTextCustom/linkResolver'

describe('Blog', () => {
    let props: BlogPageProps
    beforeEach(() => (props = blogPage))

    it('should render with a post', () => {
        const { baseElement, queryByText } = render(<Blog data={props.data} />)

        const text = queryByText('Lorem ipsum dolor sit amet')
        const date = queryByText('Jan 01, 2020')

        expect(baseElement).toMatchSnapshot()
        expect(text).toBeDefined()
        expect(date).toBeDefined()
    })

    it('clicking the title should take you to the show page', () => {
        const mockLink = jest.spyOn(Link, 'linkResolver')
        const { queryByText } = render(<Blog data={props.data} />)

        const title = queryByText('First post')

        if (title) {
            fireEvent.click(title)
        }

        expect(mockLink).toHaveBeenCalledWith({
            id: '123',
            type: 'blog_post',
            uid: 'blog-1',
        })
    })
})
