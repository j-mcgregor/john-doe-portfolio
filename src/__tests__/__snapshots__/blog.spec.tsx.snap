// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog should render with a post 1`] = `
<body>
  <div>
    <nav
      class="navbar"
    >
      <a
        class="brand"
        href="/"
      >
        John Doe
      </a>
      <ul
        class="nav-list"
      >
        <li
          class="nav-item"
        >
          <a
            href="/gallery"
          >
            Gallery
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            href="/blog"
          >
            Blog
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            href="/contact"
          >
            Contact
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="container-fluid"
    >
      <div
        class="blog pb7"
      >
        <h1>
          Blog
        </h1>
        <div
          class="post-summary"
        >
          <h2>
            <a
              href="/blog/blog-1"
            >
              First post
            </a>
          </h2>
          <p
            class="blog-post-meta"
          >
            <time>
              Jan 01, 2020
            </time>
          </p>
          <p>
            Lorem ipsum dolor sit amet
          </p>
        </div>
      </div>
    </div>
    <footer
      class="footer"
    >
      Reach me on
      <div
        class="social"
      />
    </footer>
  </div>
</body>
`;
